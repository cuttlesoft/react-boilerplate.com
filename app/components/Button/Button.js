/* eslint-disable react/prop-types */
/**
 *
 * Button.js
 *
 * A common button
 */

import React from 'react'
import { Button } from 'grommet'
import { doc } from './Button.doc'
import helpers from '../../utils/helpers'

function createButton({
  a11yTitle,
  active,
  alignSelf,
  as,
  color,
  disabled,
  fill,
  focusIndicator,
  gap,
  gridArea,
  hoverIndicator,
  href,
  icon,
  margin,
  onClick,
  plain,
  primary,
  reverse,
  target,
  title,
  type,
  ...rest
}) {
  return (
    <Button
      a11yTitle={a11yTitle}
      active={active}
      alignSelf={alignSelf}
      as={as}
      color={color}
      disabled={disabled}
      fill={fill}
      focusIndicator={focusIndicator}
      gap={gap}
      gridArea={gridArea}
      hoverIndicator={hoverIndicator}
      href={href}
      icon={icon}
      label={title}
      margin={margin}
      onClick={onClick}
      plain={plain}
      primary={primary}
      reverse={reverse}
      target={target}
      type={type}
      {...rest}
    />
  )
}

export default helpers.createWithDoc({
  envName: process.env.NODE_ENV,
  docFunction: doc,
  component: createButton,
})
