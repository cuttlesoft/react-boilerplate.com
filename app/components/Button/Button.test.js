import React from 'react'
import { fireEvent, render, cleanup } from '@testing-library/react'

import Button from './Button'

const handleRoute = () => {}
const href = 'https://cuttlesoft.com'
const children = <h1>Test</h1>
const renderComponent = (props = {}) =>
  render(
    <Button href={href} {...props}>
      {children}
    </Button>,
  )

afterEach(cleanup)

describe('Button', () => {
  it('renders a <a> tag if no route is provided', () => {
    const { container } = renderComponent({ href })

    expect(container.querySelector('a')).not.toBeNull()
  })

  it('renders a <button> tag to change route if the handleRoute prop is specified', () => {
    const { container } = renderComponent({ handleRoute })

    expect(container.querySelector('button')).toBeDefined()
  })

  it('fires of a click event', () => {
    const onClickSpy = jest.fn()
    const { container } = renderComponent({ onClick: onClickSpy })
    fireEvent.click(container.querySelector('a'))

    expect(onClickSpy).toHaveBeenCalled()
  })

  it('has children', () => {
    const { container } = renderComponent()

    expect(container.querySelector('a').children).toHaveLength(1)
  })

  it('should have a class attribute', () => {
    const { container } = renderComponent()

    expect(container.querySelector('a').hasAttribute('class')).toBe(true)
  })

  it('should not adopt a type attribute when rendering a button', () => {
    const type = 'submit'
    const { container } = renderComponent({ handleRoute, type })

    expect(container.querySelector(`button[type="${type}"]`)).toBeNull()
  })

  it('renders and matches snapshot', () => {
    const { container } = renderComponent()

    expect(container).toMatchSnapshot()
  })
})
