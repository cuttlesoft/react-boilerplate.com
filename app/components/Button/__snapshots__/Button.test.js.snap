// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders and matches snapshot 1`] = `
<div>
  <a
    class="StyledButton-sc-323bzc-0 fvdBZJ Button__StyledGrommetButton-sc-5cxguj-0 liIZVq"
    href="https://cuttlesoft.com"
  >
    <h1>
      Test
    </h1>
  </a>
</div>
`;
